Initial revision of crash application

leahnp opened the pull request:
Crash app MVP, send k2 logs and failed task name to elasticsearch cluster.
Related: 
K2 integration: https://github.com/samsung-cnct/k2/pull/505
K2-tools integration: https://github.com/samsung-cnct/k2-tools/pull/57

coffeepac commented on README.md:
```diff
@@ -1,2 +1,26 @@
+
+## License
+
+TODO: Write license
```
We use the Apache2 license for all open source projects

leahnp replied:
Added

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,81 @@
+	// check that regex found "TASK [...]", if not we assume there were no ansible
+	// tasks in the logs string parameter
+	if failedTasks == nil {
+		return "", fmt.Errorf("Log does not contain any ansible tasks")
```
I would prefer in this case to send the entire log and say, 'log does not seem to contain any ansible failed tasks'.  this is still a failure and something we want to track. 

I would guess this type of failure would be an ansible configuration failure (using a tag that doesn't exist, not specifying a hostfile that exists, etc.).

leahnp replied:
Changed the function to do this - nice call, I agree.

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,81 @@
+		return err
+	}
+
+	defer resp.Body.Close()
```
this should be placed immediately after the resp struct is created.  the defer command will ensure it gets executed as the function exits.

leahnp replied:
Moved

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,81 @@
+func send_logs(filename string) error {
+	logBuffer, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
```
for all of these 'return err' commands I would like to see you wrapping them with something like errors.Wrap("descriptive text", err) from https://github.com/pkg/errors   This will help with debugging this app to know exactly which operation is throwing a wobbly.

leahnp replied:
re-formatted error message to be more verbose with fmt.Errorf(), didn't want to bring in 3rd party package if we can avoid it. Let me know if this does the trick, original errors are still retained.

coffeepac replied:
errors is already included in this project.  I would prefer seeing errors.Wrap() instead of fmt.Errorf().

leahnp replied:
I think (confusingly) there is the standard library 'errors' package: https://golang.org/pkg/errors/ (the one I am using) and a 3rd party 'errors' package: xhttps://godoc.org/github.com/pkg/errors (the 3rd party package contains Wrap()).  Both have the same name, let me know if you still want me to import the second errors package...I wonder if that would cause name conflicts though.

coffeepac replied:
that'll teach me to not read too closely.  you're right.  leave it as-is.

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,81 @@
+	resp, err := http.DefaultClient.Do(req)
+	if (resp.StatusCode < 200 || resp.StatusCode > 299) {
+		response := fmt.Sprintf("%s", resp)
+		return errors.New(response)
```
this should also have context added to it, not sure if errors.Wrap is exactly the right thing here as its a new error but some context is great.

leahnp replied:
used  fmt.Errorf(), see above comment on subject for more clarity

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,81 @@
+		return errors.New(response)
+	}
+
+	if err != nil {
```
I don't think this expression will ever evaluate to true.  if so the whole block should be removed.

leahnp replied:
I think this would trigger is http.DefaultClient.Do(req) returned an error, I added a comment for clarity and a more verbose error message. Let me know if I am misunderstanding.

coffeepac commented on README.md:
```diff
@@ -1,2 +1,26 @@
+
+## License
+
+Apache 2.0
```
will need to include the text of the license in the repository as well.

coffeepac commented on crash-app.go:
```diff
@@ -0,0 +1,84 @@
+	"bytes"
+	"errors"
+	"regexp"
+	"net/http"
```
it came up today in channel but there should be an empty line after regexp and above net/http.  its a kubernetes/go convention.  I learned about it today.

leahnp replied:
What channel?

NathanZook commented on crash-app.go:
```diff
@@ -0,0 +1,84 @@
+		err = fmt.Errorf("Error sending POST request: %v", err)
+		return err
+	}
+
```
If `http.DefaultClient.Do` returns an error, there is no response body.  This error handling should go above the `defer resp.Body.Close()`

NathanZook commented on crash-app.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	// error handling for http.DefaultClient.Do(req) 
+	if err != nil {
```
If `http.DefaultClient.Do` returns an error, there is no response body. This error handling should go above the `defer resp.Body.Close()`.

coffeepac replied:
not technically correct, docs (somewhat opaquely) state that there will be a Body but it can be ignored (https://golang.org/pkg/net/http/#Client.Do)  

However, good point that the err handling should be done before checking the response codes

leahnp replied:
Good call, moved

leahnp commented:
Jenkinsfile is just to help me start developing on that, framework only.

leahnp commented:
I added the first pass at unit tests, let me know how you would like to see these developed. I think to  add any more tests I would need to break up my 'sendLogs' function into a few smaller functions. @coffeepac 

@NathanZook if you have time to checkout/approve if appropriate, I made the changes you suggested and would love to get this merged tonight so I can work on the Jenksinfile/CI tomorrow.

coffeepac commented:
For unit tests, that's fine.  The next step would be to either break up the sendLogs function or create a dummy http server in the unit test setUp function and have sendLogs send the logs there.  Then you could test your error handling.

But that can be left to another time if we're experiencing problems with the crash app.  Looks good!

leahnp commented:
@NathanZook I created a .gitignore, can you explain what you mean by: "(In particular, I don't believe that we want k2-crash-application to be checked in.)"

NathanZook commented:
In general, we do not check in built artifacts.  (The primary exception being for things which require special tooling.)  I was suggesting to copy the .gitignore from k2cli, but to replace k2cli in it with k2-crash-application.

leahnp commented:
@NathanZook I ran the suggested linters/formatters, changed comment style and added the .gitignore. I feel a bit overwhelmed reviewing: `Please review
https://github.com/samsung-cnct/barrel/blob/master/DEVELOPMENT.md and
https://github.com/golang/go/wiki/CodeReviewComments for changes.
(I noted issues with function commenting and test structure. Go vet catches two correctness issues.)` - it's hard to tell what applies to this project since I don't have a lot of go experience. If there is anything else outstanding you can see just let me know.

NathanZook commented on crash-app_test.go:
```diff
@@ -0,0 +1,49 @@
+	"fmt"
+	"testing"
+)
+
```
I just had a quick discussion with @coffeepac about how we want to do our tests.  K8s is our reference, so see https://github.com/kubernetes/test-infra/blob/master/kubetest/anywhere_test.go and https://github.com/kubernetes/test-infra/blob/master/prow/cmd/hook/server_test.go for a couple of examples.  If you are having difficulty, please reach out.  Hopefully, I'll be writing my own first tests in Go shortly.

Note that the names of the test are important--they communicate what is being tested.

mikeln commented on crash-app_test.go:
```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
```
Is this what we really want?  Are we saying that this is under kubernetes copyright?  shouldn't this be under "k2 authors" or similarly limited to our project(s)?

leahnp replied:
@mikeln good call. @coffeepac should we say "k2 authors" "CNCT" in the copyright?

coffeepac replied:
yep, that's what our lawyers have said.  we do what they say for legal stuff.

mikeln commented on crash-app.go:
```diff
@@ -0,0 +1,83 @@
+package main
```
Should have a copyright header.

leahnp replied:
Thanks, added

NathanZook commented on crash-app_test.go:
```diff
@@ -0,0 +1,89 @@
+		actual := tc.method(tc.arg)
+		if actual != tc.expected {
+			t.Errorf("Test '%s' failed, actual: '%s', got:  '%s'\n", tc.name, tc.expected, actual)
+		}
```
I think you mean `"Test '%s' failed, expected:  '%s', got:  '%s'\n"`.

leahnp replied:
ha, funny bug - changed it
